### PR TITLE
Add Book specs

### PR DESCRIPTION
**Pull request by Martum**
### Summary
- Add rspec and others testing gems.
- Add Book specs.

### Trello card
https://trello.com/c/v4KsWWPH/23-crear-tests-para-el-modelo-de-book

### CODE REVIEW

File: .rspec [hunk 1]
```diff
@@ -0,0 +1,2 @@
+--color
+--require spec_helper
```
Comment (francoprud) [inline .rspec]: Remove from here. This must be as a require of `rails_helper.rb` file.

File: Gemfile [hunk 2]
```diff
@@ -50,6 +50,13 @@ group :development, :test do
 
   # Static code analyzer
   gem 'rubocop', '~> 0.47.1', require: false
+  gem 'rspec-rails', '~> 3.5'
```
Comment (francoprud) [inline Gemfile]: Rspec rails inside development group? :thinking:

File: Gemfile [hunk 3]
```diff
@@ -50,6 +50,13 @@ group :development, :test do
...
+end
+
+group :test do
+  gem 'database_cleaner'
+  gem 'ffaker'
```
Comment (francoprud) [inline Gemfile]: You can use this to populate fake data for development. Send it to development and test group.

File: Gemfile [hunk 4]
```diff
@@ -50,6 +50,13 @@ group :development, :test do
...
+  gem 'factory_girl'
```
Comment (francoprud) [inline Gemfile]: You can use this to populate fake data for development. Send it to development and test group.

File: spec/models/book_spec.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
+require 'rails_helper'
+
+describe Book, type: :model do
```
Comment (francoprud) [inline spec/models/book_spec.rb]: No need for type

File: spec/models/book_spec.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  subject(:book) do
```
Comment (francoprud) [inline spec/models/book_spec.rb]: No, erase this specs and for models use the gem [shoulda-matchers](https://github.com/thoughtbot/shoulda-matchers).

To the configuration add `test_framework` for rspec and `library` rails.

File: spec/rails_helper.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+require File.expand_path('../../config/environment', __FILE__)
+# Prevent database truncation if the environment is production
+abort("The Rails environment is running in production mode!") if Rails.env.production?
+require 'spec_helper'
+require 'rspec/rails'
+# Add additional requires below this line. Rails is not loaded until this point!
+
+require 'ffaker'
+
+require 'database_cleaner'
```
Comment (francoprud) [inline spec/rails_helper.rb]: Remove `database_cleaner` config from this file and add them to `spec_helper.rb`. This doesn't depends on rails, but yes for the specs

File: spec/rails_helper.rb [hunk 8]
```diff
@@ -0,0 +1,63 @@
+# This file is copied to spec/ when you run 'rails generate rspec:install'
+ENV['RAILS_ENV'] ||= 'test'
+require File.expand_path('../../config/environment', __FILE__)
+# Prevent database truncation if the environment is production
+abort("The Rails environment is running in production mode!") if Rails.env.production?
+require 'spec_helper'
+require 'rspec/rails'
+# Add additional requires below this line. Rails is not loaded until this point!
+
+require 'ffaker'
```
Comment (francoprud) [inline spec/rails_helper.rb]: Remove this require from here and add them to `spec_helper.rb` file. This doesn't depends on rails, but yes for the specs

File: spec/spec_helper.rb [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+# individual file that may not need all of that loaded. Instead, consider making
+# a separate helper file that requires the additional dependencies and performs
+# the additional setup, and require it from the spec files that actually need
+# it.
+#
+# The `.rspec` file also contains a few flags that are not defaults but that
+# users commonly want.
+#
+# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
+RSpec.configure do |config|
```
Comment (francoprud) [inline spec/spec_helper.rb]: You are missing all the configuration from the gems.

Add the `config.include FactoryGirl::Syntax::Methods` line mentioned in the documentation.

Also add the following configuration for `database_cleaner`

```ruby
  config.before(:suite) do
    DatabaseCleaner.strategy = :transaction
    DatabaseCleaner.clean_with(:truncation)
  end

  config.around(:each) do |example|
    DatabaseCleaner.start
    DatabaseCleaner.cleaning do
      example.run
    end
  end

  config.after(:each) do
    DatabaseCleaner.clean
  end
```

**Comment by francoprud** [comment]
@Martum Travis is failing because of rubocop offenses and scss-lint. Remove scss-lint and fix that rubocp offenses

File: Gemfile (hunk 2 above)
Comment (Martum) [inline Gemfile]: The doc says it should be included inside dev and test groups. Wasn't sure why, but accordding to [this](http://stackoverflow.com/questions/10413534/why-do-people-group-rspec-under-test-and-development-in-the-gemfile) it's "... to expose generators and rake tasks without having to type RAILS_ENV=test."

Comment (francoprud) [inline Gemfile]: Nice! I prefer specifing the env when migrating env database, but you can leave it like that

File: spec/factories/book.rb [hunk 10]
```diff
@@ -0,0 +1,10 @@
+FactoryGirl.define do
+  factory :book do
+    genre FFaker::Book.genre
```
Comment (francoprud) [inline spec/factories/book.rb]: Use blocks for this and indent.

```ruby
genre     { FFaker::Book.genre }
author    { FFaker::Name.name }
image     { FFaker::Book.cover }
title     { FFaker::Book.title }
publisher { FFaker::Name.name }
year      { FFaker::Vehicle.year }

File: spec/rails_helper.rb [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+  # https://relishapp.com/rspec/rspec-rails/docs
+  config.infer_spec_type_from_file_location!
+
+  # Filter lines from Rails gems in backtraces.
+  config.filter_rails_from_backtrace!
+  # arbitrary gems may also be filtered via:
+  # config.filter_gems_from_backtrace("gem name")
+
+  # Add FactoryGirl methods
+  config.include FactoryGirl::Syntax::Methods
```
Comment (francoprud) [inline spec/rails_helper.rb]: All this configuration (factory_girl syntax and datablase_cleaner) must be inside `spec_helper.rb` file. This doesn't depends on rails, it depends on the specs.

`Shoulda::Matchers.configure` is ok!

File: spec/models/book_spec.rb [hunk 12]
```diff
@@ -0,0 +1,17 @@
+require 'rails_helper'
+
+describe Book do
+  [:genre, :author, :image, :title, :publisher, :year].each do |field|
```
Comment (francoprud) [inline spec/models/book_spec.rb]: User `%i()` for symbol array

File: spec/models/book_spec.rb [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+    it { is_expected.to validate_presence_of field }
+  end
+
+  describe '#create' do
+    context 'with valid fields' do
+      subject(:book) do
```
Comment (francoprud) [inline spec/models/book_spec.rb]: Do you need to specify the `:book` inside subject?

Can't you do the following:

```ruby
  subject { create(:book) }
```
??

**Pull request merged**